Integrating Calico as CNI provider

heschlie opened the pull request:
This PR is to add Calico as a CNI provider using `--networking calico` similar to the other CNI providers.  We use a PetSet for our own ETCD instance to avoid stepping on the main ETCD instance, and also to avoid needing to template in the cluster ETCD_ENDPOINTS.

The yaml file contains the versions which match to Calico v1.6 you can find the versions of each component [here](http://docs.projectcalico.org/v1.6/releases/)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1093)
<!-- Reviewable:end -->

heschlie commented:
I signed it!

chrislovecnm commented on hack/dev-build.sh:
```diff
@@ -65,7 +65,7 @@ MASTER_SIZE=${MASTER_SIZE:-m4.large}
 # NETWORK
 TOPOLOGY=${TOPOLOGY:-private}
-NETWORKING=${NETWORKING:-weave}
+NETWORKING=${NETWORKING:-calico}
```
Can you back this out 😀

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+# This ConfigMap is used to configure a self-hosted Calico installation.
```
So the name of this is actually causing a bug.  Can we do v1.6 or v1.6.0?  I have to fix this with weave.

heschlie replied:
Definitely

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+  namespace: kube-system
+data:
+  # The calico-etcd PetSet service IP:port
+  etcd_endpoints: "http://100.64.0.136:6666"
```
^ looks hard coded :(

heschlie replied:
It is :( see comments in service

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+
+# This PetSet lets us run a self-hosted Calico install and not interfere with the
+# cluster etcd, and not need to be aware of where it is when the cluster is coming up
+apiVersion: apps/v1alpha1
```
So this is in alpha 1.3.x and is not available in many clusters.  We need to document this, and also update this when 1.5 is out.

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+        # Run this pod on the master
+        kubernetes.io/role: master
+      hostNetwork: true
+      containers:
```
we need resource limits.

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+                fieldRef:
+                  fieldPath: status.podIP
+          command: ["/bin/sh","-c"]
+          args: ["/usr/local/bin/etcd --name=calico --data-dir=/var/etcd/calico-data --advertise-client-urls=http://$CALICO_ETCD_IP:6666 --listen-client-urls=http://0.0.0.0:6666 --listen-peer-urls=http://0.0.0.0:6667"]
```
you can do `-` with newlines instead of one line if you like.

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
```
Umm this is big enough?

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+    k8s-app: calico-etcd
+  # This ClusterIP needs to be known in advance, since we cannot rely
+  # on DNS to get access to etcd.
+  clusterIP: 100.64.0.136
```
why ... confused a tad.

caseydavenport replied:
I think @heschlie's intention here was to bypass DNS, which doesn't come up until networking is installed.

chrislovecnm replied:
how can this be hard coded?

heschlie replied:
Currently the DNS providers do not come up until the networking is online, but Calico needs to write some stuff to etcd before it comes online, which block DNS from coming up so we're unable to resolve the name of the service.  We could template in the etcd endpoints for the cluster, which would remove the need for the petset and service (and is how we normally deploy Calico) but we were asked to not use the cluster etcd.

If we were allowed to write to the cluster etcd it would remove a lot of not so great parts of this manifest.

heschlie replied:
@chrislovecnm would it be reasonable to for us to write to the cluster etcd, we don't believe Calico will step on anything, and AFAIK we have not had any reports of Calico corrupting or munging any data in etcd before.  This would help remove a lot of complexity and hardcoding in this manifest.

chrislovecnm replied:
@justinsb ??

caseydavenport replied:
Yeah.

It's something we know real Calico users do (share etcd).  Personally I'd have more concerns about the approach in this PR w.r.t the added complexity vs. just sharing with the apiserver.

chrislovecnm replied:
Well we will need to document it well in the repo. I am nervous when you talk to the source of truth directly.

Questions: to document - or point to calico docs

Etcd upgrades
Etcd ha upgrades 
Etcd version support
Backups
Recovery
Added etcd load (we may have this with weave as well)

It is up to the user to understand what they are using.  But we have an obligation to provide a production grade experience. Which I know you guys are up to do!

chrislovecnm replied:
Awesome. You guys are the experts please refactor. We are going to need some specific documentation. Including where to file issues.

Thanks

caseydavenport replied:
> We are going to need some specific documentation.

I'm sure @heschlie will be delighted to write that as part of this PR :)

Do you have a specific place in mind you'd like the Calico specific documentation?  I'm assuming that just plopping it in the existing networking doc doesn't make sense.  Really though for most things k8s projects should be directing users to project documentation rather than duplicating lots info in-tree.

@heschlie we should be:

- Referring users to projectcalico/calico to raise issues / ask for help.
- Directing users to the correct places in the Calico documentation for how to manage / upgrade / get the most out of Calico (you already have this link included in the PR it seems).
- Directing users to the [etcd documentation](https://coreos.com/etcd/docs/) for how to manage / upgrade / get the most out of etcd.

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+            - name: IP
+              value: ""
+          securityContext:
+            privileged: true
```
you need full privileged?

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+        # Runs calico/node container on each Kubernetes node.  This
+        # container programs network policy and routes on each
+        # host.
+        - name: calico-node
```
resource limits on the pod please :)

chrislovecnm commented on upup/models/cloudup/resources/addons/networking.calico/v1.6.20161206.yaml:
```diff
@@ -0,0 +1,307 @@
+          args:
+          - "pool"
+          - "add"
+          - "192.168.0.0/16"
```
hard coded?

heschlie commented:
@chrislovecnm I'm having some trouble figuring out how kops templates manifests, moreso the naming convention of the file plus where in the code I need to insure that it makes it into the KOPS_STATESTORE s3 bucket templated similar to the DNS addon.

chrislovecnm commented:
@heschlie yah the templating code is fun. We have office hours tomorrow @ 9 am Pacific if that works.

kris-nova commented:
@chrislovecnm @justinsb - we need to know if we want to try to get this into the next release - I personally think this should go in - if we can get the PR done - We need to give our overlay networks some competition ;) 

Thoughts?

chrislovecnm commented:
I do not want to shoe horn this into the next release.  We are still discussing architecture, and @justinsb and I need to come to an agreement on architecture.

heschlie commented:
I push up the requested changes.  I've also templated in the cluster etcd for now, if it turns out we do not want this I'll revert it to the petset.

caseydavenport commented on upup/models/cloudup/resources/addons/networking.calico/v2.0.yaml.template:
```diff
@@ -0,0 +1,256 @@
+# This manifest deploys the Calico policy controller on Kubernetes.
+# See https://github.com/projectcalico/k8s-policy
+apiVersion: extensions/v1beta1
+kind: ReplicaSet
```
@heschlie could you please update this to be a Deployment instead of a ReplicaSet?

chrislovecnm commented:
We need a rebase on this :)

heschlie commented:
@chrislovecnm I've added some more documentation on Calico WRT getting more info, help, and reporting issues.

chrislovecnm commented:
CI is broken :(

heschlie commented:
@chrislovecnm Fixed, didn't see the v1alpha2 added when I rebased.  I also added a couple of tests similar to what was in there for Weave, let me know if they look ok, or if I should add some more that I may have missed.

chrislovecnm commented on pkg/apis/kops/v1alpha1/networking.go:
```diff
@@ -52,3 +53,7 @@ type KopeioNetworkingSpec struct {
+
+// Calico declares that we want Calico networking
+type CalicoNetworkingSpec struct {
+}
```
Did you regenerate the api? v1alpha2 is in there https://github.com/kubernetes/kops/blob/master/docs/development/api_updates.md

chrislovecnm replied:
Ignore ... I am still learning :)

justinsb commented on upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:
```diff
@@ -131,5 +131,22 @@ func (b *BootstrapChannelBuilder) buildManifest() (*channelsapi.Addons, map[stri
 	}
 
+	if b.cluster.Spec.Networking.Calico != nil {
+		key := "networking.calico"
```
This technically should be a domain that you "own", so we don't get into naming problems.  (the weave one is not right..)

How about `networking.projectcalico.org`

chrislovecnm replied:
Guys I think this is the last change!  I reviewed with @justinsb, who is helping me learn the api stuff.

chrislovecnm commented:
Waiting on e2e

chrislovecnm commented:
Nice work guys. Now here is the ask. Please keep the yaml up to date!

heschlie commented:
@chrislovecnm You got it!  Thanks for all the help!
